File: uscore/src/test/scala/com/babylonhealth/lit/uscore/examples/Pulse_oximetryExampleTest.scala
package com.babylonhealth.lit.uscore.examples

import java.time.{ ZoneOffset, ZonedDateTime }
import java.time.temporal.ChronoUnit

import io.circe.parser.decode
import io.circe.syntax._
import org.scalatest.freespec.AnyFreeSpec
import org.scalatest.matchers.should.Matchers
import org.skyscreamer.jsonassert.JSONAssert

import com.babylonhealth.lit.core.ChoiceImplicits._
import com.babylonhealth.lit.core._
import com.babylonhealth.lit.core.model.{ CodeableConcept, Coding, Quantity, Reference, Resource }
import com.babylonhealth.lit.core.serdes.{ objectDecoder, objectEncoder }
import com.babylonhealth.lit.hl7.OBSERVATION_STATUS
import com.babylonhealth.lit.uscore.model.Us_core_pulse_oximetry

class Pulse_oximetryExampleTest extends AnyFreeSpec with Matchers {
  private val time: FHIRDateTime = FHIRDateTime(ZonedDateTime.now(ZoneOffset.ofHours(5)).truncatedTo(ChronoUnit.MILLIS))

  val example: Us_core_pulse_oximetry = Us_core_pulse_oximetry(
    code = CodeableConcept(coding = LitSeq(
      Coding(system = Some("http://loinc.org"), code = Some("2708-6")),
      Coding(system = Some("http://loinc.org"), code = Some("59408-5")))),
    status = OBSERVATION_STATUS.FINAL,
    subject = Reference(reference = Some("Patient/1234567890")),
    category = LitSeq(
      CodeableConcept(coding = LitSeq(
        Coding(
          system = Some("http://terminology.hl7.org/CodeSystem/observation-category"),
          code = Some("vital-signs"))))),
    effective = choice(time),
    value =
      Some(Quantity(system = Some("http://unitsofmeasure.org"), value = Some(94.3), code = Some("%"), unit = Some("%")))
  )

  val jsonStr: String =
    s"""{"resourceType":"Observation",
       |"meta":{"profile":["${Us_core_pulse_oximetry.profileUrl.get}"]},
       |"code":{"coding":[{"system":"http://loinc.org","code":"2708-6"},{"system":"http://loinc.org","code":"59408-5"}]},
       |"status":"final",
       |"subject":{"reference":"Patient/1234567890"},
       |"category":[{"coding":[{"system":"http://terminology.hl7.org/CodeSystem/observation-category","code":"vital-signs"}]}],
       |"effectiveDateTime":"${time.fmt}",
       |"valueQuantity":{"system":"http://unitsofmeasure.org","code":"%","unit":"%","value":94.3}}""".stripMargin

  "encodes" in {
    val exampleAsJson = example.asJson.noSpaces
    JSONAssert.assertEquals(jsonStr, exampleAsJson, true)
  }

  "decodes" in {
    val jsonAsResource = decode[Resource](jsonStr).fold(throw _, identity)
    jsonAsResource shouldEqual example
  }
}

File: uscore_java/src/test/java/com/babylonhealth/lit/uscore_java/examples/Pulse_oximetryExampleTest.java
package com.babylonhealth.lit.uscore_java.examples;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.babylonhealth.lit.core.FHIRDateTime;
import com.babylonhealth.lit.core.model.Resource;
import com.babylonhealth.lit.core_java.LitUtils;
import com.babylonhealth.lit.core_java.builders.CodeableConceptBuilder;
import com.babylonhealth.lit.core_java.builders.CodingBuilder;
import com.babylonhealth.lit.core_java.builders.QuantityBuilder;
import com.babylonhealth.lit.core_java.builders.ReferenceBuilder;
import com.babylonhealth.lit.hl7_java.codes.ObservationStatus;
import com.babylonhealth.lit.uscore.model.Us_core_pulse_oximetry;
import com.babylonhealth.lit.uscore.model.Us_core_pulse_oximetry$;
import com.babylonhealth.lit.uscore_java.builders.Us_core_pulse_oximetryBuilder;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.skyscreamer.jsonassert.JSONAssert;
import scala.math.BigDecimal;

public class Pulse_oximetryExampleTest {

  private FHIRDateTime time =
      new FHIRDateTime(
          ZonedDateTime.now(ZoneOffset.ofHours(5)).truncatedTo(ChronoUnit.MILLIS),
          LitUtils.DateTimeSpecificity.Time);

  Us_core_pulse_oximetry example =
      Us_core_pulse_oximetryBuilder.builder(
              ObservationStatus.FINAL,
              ReferenceBuilder.init().withReference("Patient/1234567890"),
              List.of(
                  CodeableConceptBuilder.init()
                      .withCoding(
                          CodingBuilder.init()
                              .withSystem(
                                  "http://terminology.hl7.org/CodeSystem/observation-category")
                              .withCode("vital-signs"))),
              Us_core_pulse_oximetryBuilder.effective(time),
              CodeableConceptBuilder.init()
                  .withCoding(
                      CodingBuilder.init().withSystem("http://loinc.org").withCode("2708-6"),
                      CodingBuilder.init().withSystem("http://loinc.org").withCode("59408-5")))
          .withValue(
              QuantityBuilder.init()
                  .withSystem("http://unitsofmeasure.org")
                  .withValue(BigDecimal.valueOf(94.3))
                  .withCode("%")
                  .withUnit("%"))
          .build();

  String jsonStr =
      "{\"resourceType\":\"Observation\","
          + String.format(
              "\"meta\":{\"profile\":[\"%s\"]},",
              Us_core_pulse_oximetry$.MODULE$.profileUrl().get())
          + "\"code\":{\"coding\":[{\"system\":\"http://loinc.org\",\"code\":\"2708-6\"},{\"system\":\"http://loinc.org\",\"code\":\"59408-5\"}]},"
          + "\"status\":\"final\","
          + "\"subject\":{\"reference\":\"Patient/1234567890\"},"
          + "\"category\":[{\"coding\":[{\"system\":\"http://terminology.hl7.org/CodeSystem/observation-category\",\"code\":\"vital-signs\"}]}],"
          + String.format("\"effectiveDateTime\":\"%s\",", time.fmt())
          + "\"valueQuantity\":{\"system\":\"http://unitsofmeasure.org\",\"code\":\"%\",\"unit\":\"%\",\"value\":94.3}}";

  @Test
  public void testEncodes() throws Exception {
    String exampleAsJson = LitUtils.encode(example);
    JSONAssert.assertEquals(jsonStr, exampleAsJson, true);
  }

  @Test
  public void testDecodes() {
    Resource jsonAsResource = LitUtils.decodeResource(jsonStr);
    assertEquals(example, jsonAsResource);
  }
}
